### DISCUSSION TITLE
SetImageColorpace or transformImageColorspace seems to have no effect on my file. Imagick converting my rgb file to gray

### DISCUSSION THREAD

**Discussion by birchcode** [ImageMagick/ImageMagick#6019] [Help] [answered]
### ImageMagick version

ImageMagick 7.1.0-59 

### Operating system, version and so on

Osx 11.6, PHP 8.2, Imagick 3.7

### Description

I am starting with a transparent image as my canvas. I can see the image has a color profile, I then load it with Imagick and write it and the saved file has lost its color profile. 

![Screenshot 2023-02-01 at 21 03 08](https://user-images.githubusercontent.com/1013599/216138822-1a79f881-2df3-44d1-b063-76ff72df0a2e.png)

Here's the stripped back code for readability:

```
// Start with our blank canvas
$im = new Imagick();
$im->readImage( 'blanksprite.png' );

// Write it
$write = $im->writeImage ( 'workingsprite.png' );
```

![Screenshot 2023-02-01 at 21 03 53](https://user-images.githubusercontent.com/1013599/216139887-37244cd1-9e3b-4313-acd2-a6fa7c913a0e.png)

As you can see, the colorspace is gone. And if I repeat the above code on the newly saved image, we also lose the alpha channel. Is this normal behaviour? 

I have tried setting colorspaces and transforming also to no avail. 

```
// Start with our blank canvas
$im = new Imagick();
$im->readImage( 'blanksprite.png' );

// Try colorspace stuff
$im->setImageColorspace(imagick::COLORSPACE_RGB); 
$im->setImageFormat("png32");
$im->transformImageColorspace(imagick::COLORSPACE_RGB);

// Write it
$write = $im->writeImage ( 'workingsprite.png' );
```

Here is blanksprite.png and the result workingsprite.png (note: they are transparent, sorry, but they are right below this line)
blanksprite.png
![blanksprite](https://user-images.githubusercontent.com/1013599/216140948-4966bb24-a787-42bf-8f45-f76c1d8f4363.png)
workingsprite.png
![workingsprite](https://user-images.githubusercontent.com/1013599/216140972-030e062a-0bb4-4db8-a31c-9c86e2d371c4.png)

I also looked at the imagick object properties to see if image magick is picking up the data and it does appear like it should be recognising the rgb:

```
date:create --- 2023-02-01T03:01:50+00:00

date:modify --- 2023-02-01T02:57:17+00:00

date:timestamp --- 2023-02-01T19:17:21+00:00

icc:copyright --- No copyright, use freely

icc:description --- sRGB IEC61966-2.1

png:iCCP --- chunk was found

png:IHDR.bit-depth-orig --- 8

png:IHDR.bit_depth --- 8

png:IHDR.color-type-orig --- 6

png:IHDR.color_type --- 6 (RGBA)

png:IHDR.interlace_method --- 0 (Not interlaced)

png:IHDR.width,height --- 8640, 450

png:pHYs --- x_res=2835, y_res=2835, units=1
```

Any ideas? Feel like this should be a very basic thing but I am stumped. Cheers!

**Comment by snibgo** [comment]
See https://imagemagick.org/script/defines.php under "png:include-chunk=value":

> As a special case, if the sRGB chunk is not excluded and the PNG encoder recognizes that the image contains the sRGB ICC profile, the PNG encoder will write the sRGB chunk instead of the entire ICC profile. To force the PNG encoder to write the sRGB profile as an iCCP chunk in the output PNG instead of the sRGB chunk, exclude the sRGB chunk.

**Reply by birchcode** [reply]
> See https://imagemagick.org/script/defines.php under "png:include-chunk=value":
> 
> > As a special case, if the sRGB chunk is not excluded and the PNG encoder recognizes that the image contains the sRGB ICC profile, the PNG encoder will write the sRGB chunk instead of the entire ICC profile. To force the PNG encoder to write the sRGB profile as an iCCP chunk in the output PNG instead of the sRGB chunk, exclude the sRGB chunk.

Greatly appreciated. 

I wasn't able to find these command line functions in the php imagick documentation so I'm not sure the've been implemented there. Do you know if these commands have been implemented?

Regardless you led me on the right path to try something else as I hadn't even considered trying working with gifs instead of png which doesn't seem to have these sRGB chunks. Fortunately simply working with gif I'll be able to move forward for now.

Thank you 🙏

**Comment by snibgo** [comment, marked as answer]
A command-line...
```
-define X=Y
```
... in Magick++ is probably (untested) done with...
```
void Magick::Image::attribute(const std::string name_,const char *value_)
```
... which calls SetImageProperty(). So I expect for PNG you would need:
```
image.attribute("png:exclude-chunk","sRGB")
```
A possible alternative:
```
image.attribute("png:preserve-iCCP","true")
```
Another possibility: use TIFF as your output format.

**Comment by birchcode** [comment]
setImageProperty worked.
```$im->setImageProperty('png:exclude-chunk','sRGB');```

We got our exclude-chunk into the properties:
```
icc:description --- sRGB IEC61966-2.1

**png:exclude-chunk --- sRGB**

png:iCCP --- chunk was found

png:IHDR.bit-depth-orig --- 8

png:IHDR.bit_depth --- 8

png:IHDR.color-type-orig --- 6

png:IHDR.color_type --- 6 (RGBA)

png:IHDR.interlace_method --- 0 (Not interlaced)

png:IHDR.width,height --- 8640, 450

png:pHYs --- x_res=2835, y_res=2835, units=1
```

And we have colour images 🥳
